### PR TITLE
feat: add Zassenhaus Hamiltonian unitary builder(#483)

### LINKED ISSUE #483: Add Zassenhaus Recursion for Hamiltonian Simulation
**Is your feature request related to a problem? Please describe.**

 QDK/Chemistry's `HamiltonianUnitaryBuilder` currently offers only Trotter-Suzuki product formulas for constructing exp(-iHt). For a target accuracy epsilon, Trotter-Suzuki of order p requires a step count whose leading error term scales as O(t^(p+1)/N^p); the prefactor is governed by nested commutators of the Hamiltonian summands but is not made explicit in the formula -- those commutators are absorbed into the step count rather than cancelled directly. For Hamiltonians whose dominant commutators are sparse or analytically tractable (lattice models with limited geometric coupling, chemistry Hamiltonians under term partitions with small inter-group commutators), this wastes steps.

 **Describe the solution you'd like**

 A new `HamiltonianUnitaryBuilder` implementation that uses a **Zassenhaus expansion** -- the dual of the Baker-Campbell-Hausdorff formula -- to construct exp(-iHt) as a product of exponentials in which low-order commutator corrections appear as explicit factors, rather than being controlled implicitly by the step count.

 The output must remain a `UnitaryRepresentation` consumable by existing downstream algorithms (e.g. `PhaseEstimation`) without changes to their interfaces.

 **Describe alternatives you've considered**

 - **Higher-order Trotter-Suzuki.** Already available; the recursive Suzuki construction has an exponentially growing step cost in the order, and its error prefactor is implicit.
 - **qDRIFT / randomized compilation.** Solves a different problem (scaling with the L1 norm of coefficients rather than with commutator structure); useful for very long Hamiltonians but does not exploit commutator sparsity.

 **Acceptance criteria**

 - A new factory registered under `HamiltonianUnitaryBuilder` (discoverable via the standard registry mechanism) accepts any `QubitHamiltonian` that the existing `"trotter"` factory accepts and returns a `UnitaryRepresentation` of the same shape (a `PauliProductFormulaContainer` with a repetition count).
 - For the 4-site open Heisenberg chain (J=1) and for the H2/STO-3G qubit Hamiltonian (Jordan-Wigner), the operator-norm error `|| U_exact - U_builder ||` evaluated at t=1.0 with a single time division decreases as O(t^(p+1)) when the configured expansion order p is varied over at least p in {2, 3, 4} -- i.e. the empirically fitted slope of log(error) vs log(t) over t in [1e-3, 1e-1] matches p+1 to within 0.1.
 - The result is consumable by `PhaseEstimation` end-to-end: estimating the ground-state energy of H2/STO-3G to within 1.6e-3 Hartree (chemical accuracy) using the new builder as the unitary source succeeds in the existing phase-estimation test harness.
 - A new pytest module under `python/tests/` covers all of the above; the full existing test suite continues to pass.
 - The new implementation is documented under "Available implementations" in `docs/source/user/comprehensive/algorithms/hamiltonian_unitary_builder.rst` in the same format as the existing `"trotter"` entry, and citations for the references below are added to `docs/source/references.bib`.

 **Additional context**

 - R. M. Wilcox, *Exponential operators and parameter differentiation in quantum physics*, J. Math. Phys. **8**, 962 (1967).
 - F. Casas, A. Murua, and M. Nadinic, *Efficient computation of the Zassenhaus formula*, Comput. Phys. Commun. **183**, 2386 (2012) [arXiv:1204.0389].
 - A. M. Childs, Y. Su, M. C. Tran, N. Wiebe, and S. Zhu, *Theory of Trotter error with commutator scaling*, Phys. Rev. X **11**, 011020 (2021) [arXiv:1912.08854].

### PR DESCRIPTION
**Pull request by GoiBasia**
Closes #483 

## Summary

The fix is mainly a Zassenhaus time-evolution path for issue #483. It does this:

- A new `"zassenhaus"` Hamiltonian unitary builder was added and registered.
- It plugs into the existing QDK/Chemistry pipeline without changing PhaseEstimation or mapper APIs.
- It outputs the same `PauliProductFormulaContainer` shape as `"trotter"`.
- `order=1` falls back to the existing first-order Trotter implementation.
- Higher orders explicitly add Zassenhaus commutator correction terms, giving expected local error scaling of `O(t^(p+1))`.
- It reuses existing Trotter error-bound logic for automatic `num_divisions` selection.
- It validates important edge cases: non-Hermitian Hamiltonians, invalid division counts, invalid error-bound settings, and tiny generated terms.
- It adds QPE evolution-time helper logic based on `phase_bound / ||H||`.
- Tests cover registry behavior, output compatibility, numerical scaling, H2/STO-3G, and PhaseEstimation chemical accuracy.
- Docs and references were updated for the new builder.

## Validation

All Python tests pass and the test file covers:

- registry availability,
- output container shape,
- order-1 Trotter fallback,
- automatic division-count behavior,
- Zassenhaus error scaling for `p = 2, 3, 4`,
- 4-site Heisenberg chain scaling,
- H2/STO-3G Jordan-Wigner scaling,
- end-to-end H2/STO-3G `PhaseEstimation` within chemical accuracy.

| System                            |   p | Expected slope |     Fitted slope |
| --------------------------------- | --: | -------------: | ---------------: |
| 4-site open Heisenberg chain, J=1 |   2 |              3 | `2.993367112344` |
| 4-site open Heisenberg chain, J=1 |   3 |              4 | `4.015929994746` |
| 4-site open Heisenberg chain, J=1 |   4 |              5 | `5.004733885337` |
| H2/STO-3G, Jordan-Wigner          |   2 |              3 | `2.999771030470` |
| H2/STO-3G, Jordan-Wigner          |   3 |              4 | `3.999999019763` |
| H2/STO-3G, Jordan-Wigner          |   4 |              5 | `4.999985332661` |

All fitted slopes are within `0.1` of the expected `p + 1`.

**PhaseEstimation Result**

Using H2/STO-3G with the new `zassenhaus` builder:

- estimated energy: `-1.852448023145` Hartree
- reference ground energy: `-1.851561658068` Hartree
- absolute error: `8.863650761604e-04` Hartree

So the PhaseEstimation result is within chemical accuracy, because `0.000886365` Hartree is below the required `0.0016` Hartree.

### CODE REVIEW

File: python/src/qdk_chemistry/algorithms/hamiltonian_unitary_builder/time_evolution/zassenhaus.py [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+        current = _poly_mul(current, _poly_exp({(i,): Fraction(1)}, order), order)
+
+    factors: list[WordPolynomial] = []
+    for degree in range(2, order + 1):
+        correction = _poly_homogeneous(_poly_add(target, current, scale=Fraction(-1)), degree)
+        factors.append(correction)
+        if correction and degree < order:
+            current = _poly_mul(current, _poly_exp(correction, order), order)
+
+    return tuple(factors)
```
[the remark this answers is not in the document]
Comment (GoiBasia) [reply python/src/qdk_chemistry/algorithms/hamiltonian_unitary_builder/time_evolution/zassenhaus.py]: Done([7dbf096](https://github.com/microsoft/qdk-chemistry/pull/535/commits/7dbf09654d36e9207707e0af13870512fada4424))
`_zassenhaus_word_factors()` now keeps `WordPolynomial` as a mutable `dict` only during internal construction, but converts the cached return value to an immutable tuple-of-items before returning. `_evaluate_word_polynomial()` was updated to iterate over those `(word, coeff)` pairs directly. I chose tuple because the downstream code only needs iteration, not mapping mutation or lookup. Since the tuple, word keys, and `Fraction` coefficients are immutable, this prevents accidental cache corruption while keeping the change minimal and tests pass.

File: python/src/qdk_chemistry/algorithms/hamiltonian_unitary_builder/time_evolution/zassenhaus.py [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+            )
+
+        return max(manual, auto)
+
+    def _hamiltonian_blocks(self, qubit_hamiltonian: QubitHamiltonian, atol: float) -> list[QubitHamiltonian]:
+        """Split the Hamiltonian into exactly exponentiable blocks."""
+        if qubit_hamiltonian.term_partition is None and len(qubit_hamiltonian.pauli_strings) > 1:
+            qubit_hamiltonian = FullCommutingTermGrouper().run(qubit_hamiltonian)
+
+        grouped = Trotter()._group_terms(qubit_hamiltonian)  # noqa: SLF001
```
[the remark this answers is not in the document]
Comment (GoiBasia) [reply python/src/qdk_chemistry/algorithms/hamiltonian_unitary_builder/time_evolution/zassenhaus.py:246]: Thanks for pointing this out. The dependency on Trotter here is intentional: previous reviewer feedback on the Zassenhaus implementations pointed to reuse the existing Trotter grouping/product-formula logic rather than duplicate it, since Trotter already contains the mature implementation for commuting grouping and decomposition.

I agree with that and would prefer to keep the current delegation to the mature Trotter and avoid reimplementing this logic in Zassenhaus. If maintainers prefer, I can follow up by extracting the shared grouping/decomposition logic into a stable helper used by both Trotter and Zassenhaus.

File: python/src/qdk_chemistry/utils/phase.py [hunk 3]
```diff
@@ -17,10 +18,56 @@
...
+    hamiltonian: _HamiltonianNormBound,
+    *,
+    phase_bound: float = np.pi,
+) -> float:
+    """Choose a QPE evolution time from the Hamiltonian norm bound.
+
+    The Hamiltonian 1-norm bounds the magnitude of every eigenvalue, so
+    ``t = phase_bound / ||H||`` keeps ``|E| t <= phase_bound`` for all
+    eigenvalues. The default ``phase_bound = π`` maximizes phase resolution
+    while staying within the principal QPE branch under that bound.
```
[the remark this answers is not in the document]
Comment (GoiBasia) [reply python/src/qdk_chemistry/utils/phase.py]: Done (https://github.com/microsoft/qdk-chemistry/pull/535/commits/7dbf09654d36e9207707e0af13870512fada4424)
I reworded the docstring to describe `schatten_norm` as a spectral-norm upper bound rather than an ambiguous “Hamiltonian 1-norm,” and clarified that `QubitHamiltonian` uses the Pauli-coefficient 1-norm, which may be loose.

File: python/src/qdk_chemistry/algorithms/hamiltonian_unitary_builder/time_evolution/zassenhaus.py [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+                continue
+
+            formula_order = _correction_formula_order(degree, order)
+            factor_terms.append(
+                Trotter(order=formula_order)._decompose_trotter_step(  # noqa: SLF001
+                    correction_hamiltonian,
+                    time=1.0,
+                    atol=atol,
+                )
+            )
```
[the remark this answers is not in the document]
Comment (GoiBasia) [reply python/src/qdk_chemistry/algorithms/hamiltonian_unitary_builder/time_evolution/zassenhaus.py:301]: Thanks, this is the same private-API concern as the grouping helper. The dependency on Trotter is intentional here: previous reviewer feedback asked Zassenhaus to reuse Trotter’s existing product-formula/decomposition logic instead of duplicating it. I also agree to keep delegating to the mature Trotter implementation, and extract these helpers into a stable shared API in a follow-up if maintainers want that cleanup.

**Comment by wavefunction91** [comment]
Close in favor of #526.
